File: python/fmvpu/bamlet/bamlet_params.py
from dataclasses import dataclass
from typing import Dict, Any
from fmvpu.amlet.amlet_params import AmletParams


@dataclass
class BamletParams:
    """Python mirror of Scala BamletParams"""
    # Number of amlet columns and rows
    n_amlet_columns: int = 2
    n_amlet_rows: int = 1
    
    # Amlet parameters
    amlet: AmletParams = None
    
    # Instruction address width
    instr_addr_width: int = 16
    
    # Instruction memory depth
    instruction_memory_depth: int = 64
    
    def __post_init__(self):
        if self.amlet is None:
            self.amlet = AmletParams()
    
    @property
    def n_amlets(self) -> int:
        return self.n_amlet_columns * self.n_amlet_rows
    
    @property
    def a_width(self) -> int:
        return self.amlet.a_width
    
    @property
    def n_loop_levels(self) -> int:
        return self.amlet.n_loop_levels
    
    # Field mapping from camelCase JSON to snake_case Python
    _FIELD_MAPPING = {
        'nAmletColumns': 'n_amlet_columns',
        'nAmletRows': 'n_amlet_rows',
        'amlet': 'amlet',
        'instrAddrWidth': 'instr_addr_width',
        'instructionMemoryDepth': 'instruction_memory_depth',
    }
    
    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> 'BamletParams':
        """Create BamletParams from dictionary with camelCase field names."""
        converted_data = {}
        for camel_key, snake_key in cls._FIELD_MAPPING.items():
            if camel_key in data:
                if camel_key == 'amlet':
                    # Convert nested amlet params
                    converted_data[snake_key] = AmletParams.from_dict(data[camel_key])
                else:
                    converted_data[snake_key] = data[camel_key]
        return cls(**converted_data)

File: src/main/scala/fmvpu/bamlet/BamletParams.scala
package fmvpu.bamlet

import chisel3._
import chisel3.util.log2Ceil
import fmvpu.amlet.AmletParams
import io.circe._
import io.circe.parser._
import io.circe.generic.auto._
import io.circe.generic.semiauto._
import scala.io.Source

case class BamletParams(
  // Number of amlet columns and rows
  nAmletColumns: Int = 2,
  nAmletRows: Int = 1,
  amlet: AmletParams,
  // Instruction memory depth
  instructionMemoryDepth: Int = 64
) {
  // Calculated parameters
  def nAmlets: Int = nAmletColumns * nAmletRows
  
  // Delegate common fields to amlet params
  def aWidth: Int = amlet.aWidth
  def nLoopLevels: Int = amlet.nLoopLevels
}


/** Companion object for BamletParams with factory methods. */
object BamletParams {
  
  // Explicit decoder for BamletParams
  implicit val BamletParamsDecoder: Decoder[BamletParams] = deriveDecoder[BamletParams]

  /** Load Bamlet parameters from a JSON configuration file.
    *
    * @param fileName Path to the JSON configuration file
    * @return BamletParams instance with configuration loaded from file
    * @throws RuntimeException if the file cannot be parsed or contains invalid parameters
    */
  def fromFile(fileName: String): BamletParams = {
    val jsonContent = Source.fromFile(fileName).mkString;
    val paramsResult = decode[BamletParams](jsonContent);
    paramsResult match {
      case Right(params) =>
        params
      case Left(error) =>
        println(s"Failed to parse JSON: ${error}")
        System.exit(1)
        null
    }
  }
}
